src/coreComponents/constitutive/permeability/SlipDependentPermeability.hpp

```diff
@@ -0,0 +1,174 @@
+
+  std::unique_ptr< ConstitutiveBase > deliverClone( string const & name,
+                                                    Group * const parent ) const override;
+
+  virtual void allocateConstitutiveData( dataRepository::Group & parent,
+                                         localIndex const numConstitutivePointsPerParentIndex ) override;
+
+  static string catalogName() { return "SlipDependentPermeability"; }
+
+  virtual string getCatalogName() const override { return catalogName(); }
+
+  /// Type of kernel wrapper for in-kernel update
+  using KernelWrapper = SlipDependentPermeabilityUpdate;
+
+  /**
+   * @brief Create an update kernel wrapper.
+   * @return the wrapper
+   */
+  KernelWrapper createKernelWrapper() const
+  {
+    return KernelWrapper( m_permeability,
+                          m_dPerm_dPressure,
+                          m_dPerm_dDispJump,
+                          m_shearDispThreshold,
+                          m_maxPermMultiplier,
+                          m_initialPermeability );
+  }
+
+  struct viewKeyStruct : public PermeabilityBase::viewKeyStruct
+  {
+    static constexpr char const * dPerm_dDispJumpString() { return "dPerm_dDispJump"; }
+    static constexpr char const * shearDispThresholdString() { return "shearDispThreshold"; }
+    static constexpr char const * maxPermMultiplierString() { return "maxPermMultiplier"; }
+    static constexpr char const * iniPermeabilityString() { return "iniPermeability"; }
+  } ;
+
+private:
+
+  /// Derivative of fracture permeability to shear displacement jump between fracture surfaces
+  array4d< real64 > m_dPerm_dDispJump;
+
+  /// Threshold of shear displacement
+  real64 m_shearDispThreshold;
+
+  /// Maximum permeability multiplier
+  real64 m_maxPermMultiplier;
+
+  /// Initial permeability tensor
+  array3d< real64 > m_initialPermeability;
+
+};
+
+
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+void SlipDependentPermeabilityUpdate::compute( real64 const ( &dispJump )[3],
+                                               arraySlice1d< real64 > const & initialPermeability,
+                                               arraySlice1d< real64 > const & permeability,
+                                               arraySlice2d< real64 > const & dPerm_dDispJump ) const
+{ 
+  real64 const shearMag = std::sqrt( dispJump[1]*dispJump[1] + dispJump[2]*dispJump[2] );
+  
+  real64 const tmpTanh = std::tanh ( 3.0 * shearMag/m_shearDispThreshold );
+  
+  real64 const permMultiplier = ( m_maxPermMultiplier - 1.0 ) * tmpTanh + 1.0;
+
+  real64 const dpermMultiplier_dshearMag = ( m_maxPermMultiplier - 1.0 ) * ( 1.0 - tmpTanh * tmpTanh ) * 3.0/m_shearDispThreshold;  
+ 
+  for( localIndex i=0; i < permeability.size(); i++ )
+  {
+    permeability[i] = permMultiplier * initialPermeability[i];
+    dPerm_dDispJump[i][0] = 0.0;
+    real64 const tmpValue = initialPermeability[i] * dpermMultiplier_dshearMag /shearMag;
```

Comment:
@jhuang2601 the issue is this division by shearMag whenver the tangential components of dispJump are 0. We are trying to divide by 0 which is obviously a problem. Can you, 
1) check that all derivatives are correct. 
2) think of what the derivative value should be for whenver the shearMag is 0.